### ISSUE TITLE
Support for GOTO input

### ISSUE THREAD

**Issue by rafaelsamenezes** [esbmc/esbmc#557]
How is ESBMC behaving when dealing with GOTO input, from:

1. Our own c2goto, can we use ESBMC to export a GOTO program to be used later?
2. From CBMC, can we use the GOTO generated by/for it?

**Comment by fbrausse** [comment]
c2goto patched (in order to also export the `__ESBMC_main` symbol) with
```diff
diff --git a/src/c2goto/c2goto.cpp b/src/c2goto/c2goto.cpp
index 5aa2dbb7c..c4e1dfebc 100644
--- a/src/c2goto/c2goto.cpp
+++ b/src/c2goto/c2goto.cpp
@@ -65,6 +65,8 @@ public:
       return 1;
     if(typecheck())
       return 1;
+    if(language_files.final(context))
+      return 1;
 
     std::ofstream out(
       cmdline.getval("output"), std::ios::out | std::ios::binary);
```
generates .goto files which can be read by `esmbc --binary`. However, due to c2goto serializing an empty goto-functions table (see below), esbmc fails on such inputs with
```
GOTO instruction type NO_INSTRUCTION_TYPE not handled in goto_symext::symex_step
```

Regarding CBMC, there are a few differences:

ESBMC's serialization format consists of 4 parts:
1. header `GBF`
2. GOTO-serialization version number 1
3. symbol table
4. goto-functions table

CBMC serialization:
1. headers `GBF`, `\x7fGBF`, `\x7fELF`, various other formats like ELF, Mach-O, ...
2. GOTO-serialization version number 5
3. symbol table
4. goto-functions table

I have not checked what differences there are regarding 3. and 4.

**Comment by rafaelsamenezes** [comment]
Thanks for the report @fbrausse!

> c2goto patched (in order to also export the __ESBMC_main symbol) with

I don't think that the __ESBMC_main symbol is needed, as it can change based on the analysis type.

> I have not checked what differences there are regarding 3. and 4.

This will probably be the trick part.

**Comment by fbrausse** [comment]
> This will probably be the trick part.

I agree. Is the CBMC commit known where ESBMC forked? That would make tracking these changes easier.

**Comment by mikhailramalho** [comment]
On Sun, 2 Jan 2022, 10:02 Franz Brauße, ***@***.***> wrote:

> This will probably be the trick part.
>
> I agree. Is the CBMC commit known where ESBMC forked? That would make
> tracking these changes easier.
>

I tried to track that once but when esbmc was forked, cbmc was using CVS
and Lucas only received a dump of the tree, with no history.

Also, the first four years of esbmc development was done by using the
zip-the-source-code-daily technique, so we have no history about that
either :/

**Comment by fbrausse** [comment]
This is the differences in the GOTO instructions known to CBMC vs. ESBMC:
```diff
--- cbmc/src/goto-programs/goto_program.h       2021-11-17 16:27:29.665617442 +0100
+++ esbmc/src/goto-programs/goto_program.h      2022-01-19 19:10:59.219741997 +0100
@@
-/// The type of an instruction in a GOTO program.
-enum goto_program_instruction_typet
+typedef enum
 {
   NO_INSTRUCTION_TYPE = 0,
   GOTO = 1,              // branch, possibly guarded
   ASSUME = 2,            // non-failing guarded self loop
   ASSERT = 3,            // assertions
   OTHER = 4,             // anything else
   SKIP = 5,              // just advance the PC
-  START_THREAD = 6,      // spawns an asynchronous thread
-  END_THREAD = 7,        // end the current thread
   LOCATION = 8,          // semantically like SKIP
   END_FUNCTION = 9,      // exit point of a function
   ATOMIC_BEGIN = 10,     // marks a block without interleavings
   ATOMIC_END = 11,       // end of a block without interleavings
-  SET_RETURN_VALUE = 12, // set function return value (no control-flow change)
+  RETURN = 12,        // return from a function
   ASSIGN = 13,           // assignment lhs:=rhs
   DECL = 14,             // declare a local variable
   DEAD = 15,             // marks the end-of-live of a local variable
   FUNCTION_CALL = 16,    // call a function
   THROW = 17,            // throw an exception
-  CATCH = 18,            // push, pop or enter an exception handler
-  INCOMPLETE_GOTO = 19   // goto where target is yet to be determined
-};
+  CATCH = 18,         // catch an exception
+  THROW_DECL = 19,    // list of throws that a function can throw
+  THROW_DECL_END = 20 // end of throw declaration
+} goto_program_instruction_typet;
```
Obviously instruction 12 ~seems unchanged except for the name, needs confirmation~ _edit:_ has changed - can we assign it a new value? Other questions before we can resolve the differences between symbol- and goto-functions-tables:
- `CATCH` seems to do more in CBMC than in ESBMC.
- how exactly are the `THROW_DECL*` used in ESBMC? Is there a corresponding way to handle those in CBMC, e.g., via `CATCH`? Can we re-assign unused values (e.g., 20,21) to these in order to not overlap CBMC's values?
- does ESBMC have an equivalent of `INCOMPLETE_GOTO`?
- what does CBMC do differently with threads?

**Comment by lucasccordeiro** [comment]
> CATCH seems to do more in CBMC than in ESBMC.
> how exactly are the THROW_DECL* used in ESBMC? Is there a corresponding way to handle those in CBMC, e.g., via CATCH? Can we re-assign unused values (e.g., 20,21) to these in order to not overlap CBMC's values?

@fbrausse: This paper describes what we do with the `CATCH` and `THROW_DECL*` GOTO instructions:

https://onlinelibrary.wiley.com/doi/epdf/10.1002/stvr.1793

**Comment by lucasccordeiro** [comment]
> what does CBMC do differently with threads?

CBMC's concurrency support is an entirely symbolic encoding of a concurrent program in one SAT formulae, while ESBMC explores each interleaving individually.

**Comment by lucasccordeiro** [comment]
> does ESBMC have an equivalent of INCOMPLETE_GOTO?

As far as know, no.

**Comment by fbrausse** [comment]
>> what does CBMC do differently with threads?

> CBMC's concurrency support is an entirely symbolic encoding of a concurrent program in one SAT formulae, while ESBMC explores each interleaving individually.

I meant in terms of representation in the GOTO language.

**Comment by lucasccordeiro** [comment]
We use intrinsic functions that replace the `START_THREAD` and `END_THREAD` of CBMC. For example:

https://github.com/esbmc/esbmc/blob/master/src/goto-symex/symex_main.cpp#L416
https://github.com/esbmc/esbmc/blob/master/src/c2goto/library/pthread_lib.c#L201
https://github.com/esbmc/esbmc/blob/master/src/c2goto/library/pthread_lib.c#L211
https://github.com/esbmc/esbmc/blob/master/src/c2goto/library/pthread_lib.c#L166
https://github.com/esbmc/esbmc/blob/master/src/c2goto/library/pthread_lib.c#L120
https://github.com/esbmc/esbmc/blob/master/src/c2goto/library/pthread_lib.c#L128

**Comment by fbrausse** [comment]
Many thanks @lucasccordeiro!
> From CBMC, can we use the GOTO generated by/for it?

From Lucas' explanations it seems the languages have diverged quite substantially. It may be easier and less fragile to treat CBMC's as a new language and translate it to ESBMC rather than merging both of them.

**Comment by mikhailramalho** [comment]
> * how exactly are the `THROW_DECL*` used in ESBMC? Is there a corresponding way to handle those in CBMC, e.g., via `CATCH`? Can we re-assign unused values (e.g., 20,21) to these in order to not overlap CBMC's values?

Btw, we can probably get rid of the throw_decl instructions and model exception declarations to the function frame!

> * does ESBMC have an equivalent of `INCOMPLETE_GOTO`?

I don't think so but that shouldn't be hard to implement. I also remember label declarations being disabled at some point.

**Comment by rafaelsamenezes** [comment]
This is implemented in a separate project (goto-transcoder).
